### PR TITLE
Convert job_manager to be async

### PR DESCRIPTION
**Pull request by alanwguo**
Updates jobs api
Updates snapshot api
Updates state api

Increases jobs api version to 2

Signed-off-by: Alan Guo <aguo@anyscale.com>

## Why are these changes needed?
follow-up for https://github.com/ray-project/ray/pull/25902#discussion_r930299122

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
nice! I will review this today

**Review by architkulkarni — APPROVED** [review]
Looks good to me, thanks for the quick followup!

File: dashboard/modules/snapshot/snapshot_head.py [hunk 1]
```diff
@@ -398,6 +401,10 @@ async def run(self, server):
         self._gcs_actor_info_stub = gcs_service_pb2_grpc.ActorInfoGcsServiceStub(
             self._dashboard_head.aiogrpc_gcs_channel
         )
+        if not self._job_info_client:
+            self._job_info_client = JobInfoStorageClient(
+                self._dashboard_head.gcs_aio_client
+            )
```
Comment (edoakes) [inline dashboard/modules/snapshot/snapshot_head.py:407]: why do we lazily construct this? add comment for future reader

File: dashboard/modules/version.py [hunk 2]
```diff
@@ -1,7 +1,10 @@
 from dataclasses import dataclass
 
 # Version 0 -> 1: Added log streaming and changed behavior of job logs cli.
-CURRENT_VERSION = "1"
+# Version 1 -> 2: - Renamed job_id to submission_id.
+#                 - Changed list_jobs sdk/cli/api to return a list
+#                   instead of a dictionary.
+CURRENT_VERSION = "2"
```
Comment (edoakes) [inline dashboard/modules/version.py:7]: is this intended? or an artifact of stacked PRs? please make this a separate PR either way.

Comment (alanwguo) [reply dashboard/modules/version.py:7]: Intended but yes, an orthogonal change. Will separate.

Comment (edoakes) [reply dashboard/modules/version.py:7]: Thanks. LGTM to merge after this is resolved.

**Comment by rkooo567** [comment]
I will take a look at the PR tomorrow

**Review by edoakes — APPROVED** [review]

File: dashboard/modules/job/common.py [hunk 3]
```diff
@@ -97,30 +95,34 @@ class JobInfoStorageClient:
...
+    def __init__(self, gcs_aio_client: GcsAioClient):
+        self._gcs_aio_client = gcs_aio_client
         assert _internal_kv_initialized()
 
-    def put_info(self, job_id: str, data: JobInfo):
-        _internal_kv_put(
-            self.JOB_DATA_KEY.format(job_id=job_id),
+    async def put_info(self, job_id: str, data: JobInfo):
+        await self._gcs_aio_client.internal_kv_put(
+            self.JOB_DATA_KEY.format(job_id=job_id).encode(),
```
Comment (rkooo567) [inline dashboard/modules/job/common.py:104]: Btw, do you know what happens if you don't specify `encode `? Does it just fail?

File: dashboard/modules/job/common.py [hunk 4]
```diff
@@ -156,7 +158,7 @@ def get_all_jobs(self) -> Dict[str, JobInfo]:
                 self.JOB_DATA_KEY_PREFIX
             ), "Unexpected format for internal_kv key for Job submission"
             job_ids.append(job_id_with_prefix[len(self.JOB_DATA_KEY_PREFIX) :])
-        return {job_id: self.get_info(job_id) for job_id in job_ids}
+        return {job_id: await self.get_info(job_id) for job_id in job_ids}
```
Comment (rkooo567) [inline dashboard/modules/job/common.py:161]: This will be blocking calls right? Maybe we should use asyncio.gather here?

**Review by rkooo567 — CHANGES_REQUESTED** [review]
One comment to handle... Other things LGTM.

File: dashboard/modules/job/common.py (hunk 4 above)
Comment (alanwguo) [reply dashboard/modules/job/common.py]: nice catch

**Review by rkooo567 — APPROVED** [review]

File: dashboard/modules/job/common.py [hunk 5]
```diff
@@ -156,7 +159,17 @@ def get_all_jobs(self) -> Dict[str, JobInfo]:
...
-        return {job_id: self.get_info(job_id) for job_id in job_ids}
+
+        async def get_job_info(job_id: str):
+            job_info = await self.get_info(job_id)
+            return job_id, job_info
+
+        return {
+            job_id: job_info
+            for job_id, job_info in await asyncio.gather(
+                *[get_job_info(job_id) for job_id in job_ids]
```
Comment (rkooo567) [inline dashboard/modules/job/common.py:170]: Btw, this is not a blocker for the PR. But I wonder @iycheng do we have APIs to get a list of entities from the internal kv?

**Comment by alanwguo** [comment]
@rkooo567 @architkulkarni @edoakes , please merge!
